README.md
# <i class="icon-file"></i> [![R Project][logo]] (https://www.r-project.org/) 

Codigos de `R` utilizados para:

> - Shiny
> - Markdown
> - scripts
> - Dashboards
> - Conexiones (mongoDB)

*Mis comentarios son suposiciones relacionadas a lo que he aprendido*
[logo]: img/Rlogo.png



CHANGELOG
git log v2.1.0...v2.1.1 --pretty=format:'<li> <a href="http://github.com/jerel/<project>/commit/%H">view commit &bull;</a> %s</li> ' --reverse | grep "#changelog"

Shiny/Calcus_Impuestos/ui.R
library(shiny)

Tamano_Barras <- 300

## SideBar ----
Sidebar <- dashboardSidebar(
## Tamaño Sidebar ----
  width = Tamano_Barras-100,
## SideBar items ----
  sidebarMenu(
    menuItem("Retencion ISR", tabName = "ISR_Calculo", icon = icon("hand-spock-o"))
    , menuItem("Retencion IMSS", tabName = "IMSS_Calculo", icon = icon("hand-spock-o"))
    , menuItem("Salario", tabName = "Salario_Calculo", icon = icon("hand-spock-o"))
    # , menuItem("Optimizador", tabName = "Optimizador_Calculo", icon = icon("exchange"))
    # , menuItem("Pruebas", tabName = "Pruebas", icon = icon("spinner"))
  )
)

## Body ----
Body <- dashboardBody(
## Body Items ----
  tabItems(
## Body Salario ----
    tabItem(tabName = "ISR_Calculo",
            ## Esta seccion solo funciona con los strings predefinidos
            ## primary, success, info, warning y se cambia de acuerdo a la necesidad
            tags$style(HTML("
                  .box.box-solid.box-info>.box-header {
                    color:#F72C00;background:#F72C00}
                  .box.box-solid.box-info{
                    border-bottom-color:#3ED4C8;
                    border-left-color:#3ED4C8;
                    border-right-color:#3ED4C8;
                    border-top-color:#3ED4C8;}")),
            h1("Ingresa salario"),
            fluidRow(width = 5
              , box(width = 5, title = "Monto de ISR", status = "primary"
                     , color = "blue", solidHeader = TRUE, background = "navy"
                     , fluidRow(width = 12, uiOutput("ISR_Anual"))
                     , fluidRow(width = 12, uiOutput("ISR_MenQuin"))
                     )
              , box(width = 6, title = "Caracteristicas", status = "primary"
                  , color = "blue", solidHeader = TRUE, background = "navy",
                selectInput(inputId = "tSalario"
                            , label = "Tipo de Salario (Quincenal-Mensual)"
                            , choices = c("Quincenal", "Mensual")
                            , selected = "Mensual"),
                numericInput(inputId = "Salario", label = "Salario del periodo", min = 1
                             , step = 500, value = 22733.23)
                  )
                )
            , fluidRow(width = 12,
              # column(width = 6,
              box(width = 5, title = "Mas detalles", status = "primary"
                  , color = "blue", solidHeader = TRUE, background = "navy"
                  , collapsible = TRUE, collapsed = TRUE
                  , fluidRow(width = 12, uiOutput("oLimiteInferior"))
                  , fluidRow(width = 12, uiOutput("oExcedente"))
                  , fluidRow(width = 12, uiOutput("oCuotaFija"))
                )
              # )
            )
            # , fluidRow(width = 12,
            #          tags$head(tags$style(HTML('.small-box {min-height: 25px;} 
            #                                    .small-box-icon {height: 25px; line-height: 25px;} 
            #                                    .small-box-content {padding-top: 0px; padding-bottom: 0px;}'))),
            #   box(width = 8, solidHeader = TRUE, 
            #     box(width = 8, background = "blue", height = 300,
            #         "jola"#uiOutput("Prueba3")
            #         )
            #       
            #     )
            #   )
    ),
## Body IMSS ----
tabItem(tabName = "IMSS_Calculo",
        tags$style(HTML("
                  .box.box-solid.box-info>.box-header {
                    color:#F72C00;background:#F72C00}
                  .box.box-solid.box-info{
                    border-bottom-color:#3ED4C8;
                    border-left-color:#3ED4C8;
                    border-right-color:#3ED4C8;
                    border-top-color:#3ED4C8;}")),
        fluidRow(width = 6,
                   box(width = 8, status = "primary" #, collapsible = TRUE, collapsed = TRUE
                       , title = "Descripción", solidHeader = TRUE
                       , h4("Para el cálculo del Salario Diario Integrado (SDI) es necesario ingresar las 
                        distintas prestaciones que ofrece tu empresa."))
                )
        , fluidRow(width = 12
             , column(width = 5
                          , box(width = 12, title = "Retencion IMSS", solidHeader = TRUE, background = "navy"
                                , status = "primary", color = "blue"
                                , uiOutput("oIMSSRet"))
                          , box(width = 12, title = "Detalles", solidHeader = TRUE
                                , background = "navy", collapsible = TRUE, collapsed = TRUE
                                , fluidRow(width = 6
                                           , uiOutput("oSDI")
                                           , uiOutput("oEsEx")
                                           , uiOutput("oPrestaciones")
                                           , uiOutput("oPensiones")
                                           , uiOutput("oInvalidez")
                                           , uiOutput("oCesantia")
                                           , uiOutput("oIMSSMens")
                                           , uiOutput("oCesantiaB")
                                )
                          )
                 )
        , fluidRow(width = 6
                        , column(width = 6
                                 , div(class = "well well-sm"
                                       ,style = "background-color: #3c8dbc;color: white"
                                       ,h4("Ingrese prestaciones:")
                                 )
                        )
        , column(width =3
            , fluidRow(width = 12, 
              box(width = 12, solidHeader = TRUE, status = "info", background = "light-blue"#, height = 485
                  , selectInput("MesCons", label = "Selecciona mes de consulta.",
                                choices = sMeses, selected = format(Sys.Date(), format = "%B"))#format(Sys.Date(), format = "%B")
                  , numericInput(inputId = "SalMens" , label = "Salario Mensual(Bruto)"
                              , value = 22733.23, step = 500)
                  , numericInput(inputId = "Antiguedad" , label = "Antigüedad (Años)"
                                 , value = 1, step = 1)
                  , numericInput(inputId = "PrimVac" , label = "Prima Vacacional (%)"
                                 , value = 25, step = 1)
                  # , numericInput(inputId = "Aguinaldo" , label = "Aguinaldo (dias)"
                  #                , value = 15, step = 5)
                    )
                  )
                )
                  # , includeCSS("www/Box_Style.css")
        , column(width = 3
            , fluidRow(width = 12
                  , box(width = 12, solidHeader = TRUE, status = "info", background = "light-blue"
                  # , numericInput(inputId = "SMV", label = "Salario Minimo Vigente DF"
                  #                      , value = 73.04, min = 73.04, step = 2)
                  , numericInput(inputId = "Aguinaldo" , label = "Aguinaldo (dias)"
                                 , value = 15, step = 5)
                  , tags$div(class = "well well-sm", style = "background-color: #3c8dbc;"
                                  , "Vacaciones por ley:"
                                  , tags$h4(textOutput("oDiasVac") #textOutput("oDiasVac")
                                            , align = "center"))
                  , tags$div(class = "well well-sm", style = "background-color: #3c8dbc;"
                             , "Salario Minimo Vigente:"
                             , tags$h4(textOutput("oSMV")
                                       , align = "center"))
                  , checkboxInput(inputId = "CusDiasVac", label = "Modificar vacaciones", value = FALSE)
                  , conditionalPanel(
                          condition = "input.CusDiasVac == true",
                          numericInput(inputId = "DiasVacMod", label = "Dias de Vacaciones", value = 6)
                    )
                  )
                )
              )
            )
          )
        ),
## Body Optimizador ----
    tabItem(tabName = "Optimizador",
            h2("¿Cuanto dinero neto quieres ganar?"),
            tags$div(class = "info-box bg-red")),
    tabItem(tabName = "Pruebas",
        h2("¿Cuanto dinero neto quieres tener?"),
        tags$div(class = "info-box bg-red"),
        box(width = 5
            , tags$div(class="info-box bg-light-blue", checked=NA
                       , tags$span(class = "info-box-icon", icon("minus-square")),
                       tags$div(class="info-box-content",
                                tags$span(class ="info-box-text", "Likes"),
                                tags$span(class = "info-box-number","41400"),
                                tags$div(class = "progress",
                                         tags$div(class = "progress-bar", style="width: 70%")),
                                tags$span(class = "progress-description", "70% Increase in 30 Days")
                       )
                    )
                  )
            # ,<div style="clear:both">
            #   <a href="http://www.visualcapitalist.com/chart-epic-collapse-deutsche-bank/">
            #   <img src="http://2oqz471sa19h3vbwa53m33yj.wpengine.netdna-cdn.com/wp-content/uploads/2016/07/deutsche-bank-fall-chart.png" border="0" />
            #   </a>
            #   </div>
            #   <div>Courtesy of: <a href="http://www.visualcapitalist.com">Visual Capitalist</a></div>
            , tags$div(style = "clear:both"
                       , tags$a(
                         href = "visualcapitalist.com/chart-epic-collapse-deutsche-bank/",
                         tags$img(src = paste0("http://2oqz471sa19h3vbwa53m33yj.wpengine.netdna-cdn.com/wp-content/"
                                               ,"uploads/2016/07/deutsche-bank-fall-chart.png",sep = "")
                                  , width = "600px", height = "450px")
                            )
                          )
            # , progressBox(title = "Prueba", iconBx = "minus-square"
            #               , value = 50, barStatus = 50, description = "subtitulo")
        )
  )
)

# Dashboard Page ----
ui <- dashboardPage(skin = "blue",
                    dashboardHeader(title = "Calculadora de Impuestos", titleWidth = Tamano_Barras),
                    Sidebar,
                    Body
)


Shiny/Calcus_Impuestos/global.R
# Calculadora que permite tener un aproximado bastante real
# al calculo de ISR e IMSS
# Informacion actualizada al 2016
# Fuente ISR: http://www.sat.gob.mx/informacion_fiscal/tablas_indicadores/Paginas/tarifas_pp.aspx

library(shiny)
library(shinydashboard)
# library(DT)
library(dplyr)

# source("progressBox.R")

## Cargo informacion de ISR ----
# ISR <- readRDS(paste(dirname(rstudioapi::getActiveDocumentContext()$path),
#                       "/Data/ISR_Limites.Rds",sep = ""))

ISR <- readRDS(paste0("Data/ISR_Limites.Rds",sep = ""))

## Cargo informacion de IMSS ----
# IMSS <- readRDS(paste(dirname(rstudioapi::getActiveDocumentContext()$path),
#                      "/Data/IMSS_Porcentajes.Rds",sep = ""))

IMSS <- readRDS(paste0("Data/IMSS_Porcentajes.Rds",sep = ""))

# IMSS <- readRDS(paste("C:/Users/rafael.barragan/Documents/Codigos/Codes/Shiny/",
#                       "Calcus_Impuestos/Data/IMSS_Porcentajes.Rds",sep = ""))

## Listado de deducciones ----
ListaDeducciones <- c("Limite Inferior", "Excedente", "% Excedente"
                      , "Cuota Fija", "ISR Anual", "ISR Mensual"
                      , "ISR Quincenal", "Salario Neto Mensual"
                      , "Salario Neto Quincenal")
## String Meses ----
sMeses <- format(seq(as.Date("2000/1/1"), by = "month", length.out = 12), format = "%B")

## Funcion que calcula el salario ----
fISRAnual <- function(LimInf, CuotaFija, Aplicable, SalarioBrutoAnual){
  # Calculo Excedente al cual se le aplicara impuestos
  Excedente <- SalarioBrutoAnual - LimInf
  fISRAnual <- CuotaFija + Excedente*Aplicable
}

# Progress Box ----
progressBox <- function(title, value = 0, leftColor = "navy", rightColor = NULL, iconBx = "lightbulb-o"
                        , barStatus = 0, barColor = "light-blue", description){
  tags$div(class= paste0("info-box", if(!is.null(rightColor)) paste0(" bg-",rightColor)), checked=NA
           , tags$span(class = paste0("info-box-icon",if(!is.null(leftColor)) paste0(" bg-", leftColor))
                       , icon(iconBx)),
           tags$div(class="info-box-content",
                    tags$span(class ="info-box-text", title),
                    tags$span(class = "info-box-number",value),
                    tags$div(class = "progress",
                         tags$div(class = paste0("progress-bar", 
                                             if(is.null(rightColor)) paste0(" bg-", barColor))
                                             , style=paste0("width: ",barStatus,"%"))),
                    tags$span(class = "progress-description", description)
        )
      )
    }

## Copio desde Excel ----
# paste.table <- function() {
#   f <- file(description = 'clipboard', open = 'r')
#   df <- read.table(f, sep = '\t', header = TRUE)
#   close(f)
#   return(df)
# }

# ISR <- paste.table()
# saveRDS(ISR,"C:/Users/rafael.barragan/Documents/Codigos/Codes/Shiny/Calcus_Impuestos/Data/ISR_Limites.Rds")
# 
# IMSS <- paste.table()
# saveRDS(IMSS,"C:/Users/rafael.barragan/Documents/Codigos/Codes/Shiny/Calcus_Impuestos/Data/IMSS_Porcentajes.Rds")


Shiny/Calcus_Impuestos/server.R
library(shiny)

server <- shinyServer(function(input, output) {

## Salario: Tipo de Salario ----
  TipoSal <- reactive({
                TipoSalario <- switch(input$tSalario,
                                    "Mensual" = 1,
                                    "Quincenal" = 2)
                
                TipoSalario})
  
## Salario: Salario Anual ----  
  SalarioBaseAnual <- reactive({
    SalBAnual <- TipoSal()*input$Salario*12
  })

## Salario: Limite Inferio ----
  LimInferior <- reactive({
    
    SalBAnual <- SalarioBaseAnual()
    
    Lim_Inf <- ISR %>%
      filter(Limite_Inferior <= SalBAnual,
             Limite_Superior >= SalBAnual) %>%
      select(Limite_Inferior)
    
    Lim_Inf
  })

## Salario: Cuota Fija ----  
  CuotaFija <- reactive({
    
    SalBAnual <- SalarioBaseAnual()
    
    Cuota_fija <- ISR %>%
      filter(Limite_Inferior <= SalBAnual,
             Limite_Superior >= SalBAnual) %>%
      select(Cuota_fija)
    
    Cuota_fija
  })
  
## Salario: Aplicable ----
  Aplicable <- reactive({
    
    SalBAnual <- SalarioBaseAnual()
    
    Aplicable <- ISR %>%
      filter(Limite_Inferior <= SalBAnual,
             Limite_Superior >= SalBAnual) %>%
      select(Aplicable)
    
    Aplicable
  })

## Salario: Excedente ----  
  Excedente <- reactive({
      Excedente <- SalarioBaseAnual() - LimInferior()
    })

## IMSS: Salario Minimo ----
  SMV <- 73.04
## IMSS: Vacaciones por Ley ----
  VPL <- reactive({
    if (input$Antiguedad <5)
      {DV <- 2*input$Antiguedad+4}
    else
      {DV <- 12 + as.integer(input$Antiguedad/5)*2}
    
    DV
  })
## IMSS: Dias del Mes ----
  DiasMes <- reactive({
      ## Obtengo el Mes y año de consulta
      DM <- which(sMeses %in% input$MesCons)
      AnoConsulta <- format(Sys.Date(),"%Y")
      ## Obtengo los dias del mes de la consulta
      Dias <- format(as.Date(paste0("01/",(DM %% 12)+1,"/", AnoConsulta),"%d/%m/%Y")-1,"%d")
    })

## IMSS: SDI ----
  SDI <- reactive({
    
    SalMaxApl <- SMV*25
    
    if(input$CusDiasVac == TRUE)
      {Vacaciones <- input$DiasVacMod}
    else
      {Vacaciones <- VPL()}
    
    FactorSinTrunc <- (((Vacaciones*input$PrimVac/100)+input$Aguinaldo)/365)+1
    FactorTruncado <- trunc(FactorSinTrunc*10000)/10000
    
    SalMensual <- input$SalMens
    DM <- as.numeric(DiasMes())
    
    SalarioXFact <- FactorTruncado * SalMensual/DM
      
    if (SalarioXFact < SalMaxApl)
      {SalDI <- SalarioXFact}
    else
      {SalDI <- SalMaxApl}
    
    SalDITruncado <- trunc(SalDI*100)/100
  })

## IMSS: Especie Excedente ----
  Espec_Exed <- reactive({
    
    ## Obtengo la informacion de las retenciones del IMSS
    Retenciones_IMSS <- IMSS %>%
      filter(Enfermedades_Maternidad2 == "Especie_Excedente") %>%
      select(Empleado)
    ## Valido la informacion
    if(SDI() > 3*SMV)
      {Esp_Ex <- Retenciones_IMSS}
    else
      {Esp_Ex <- 0}
    
    DM <- as.numeric(DiasMes())
    
    Retencion <- Esp_Ex/100 * (SDI()-(3*SMV)) * DM
    
  })

## IMSS: Prestaciones en Dinero ----
  Prestaciones <- reactive({
    
    ## Obtengo la informacion de las retenciones del IMSS
    Retenciones_IMSS <- IMSS %>%
      filter(Enfermedades_Maternidad2 == "Prestaciones") %>%
      select(Empleado)
    
    DM <- as.numeric(DiasMes())
    
    Retencion <- Retenciones_IMSS/100 * SDI() * DM
    
  })

## IMSS: Pensiones y Beneficiarios ----
  Pensiones <- reactive({
    
    ## Obtengo la informacion de las retenciones del IMSS
    Retenciones_IMSS <- IMSS %>%
      filter(Enfermedades_Maternidad2 == "Pensionados_y_Beneficiarios") %>%
      select(Empleado)
    
    DM <- as.numeric(DiasMes())
    
    Retencion <- Retenciones_IMSS/100 * SDI() * DM
    
  })
  
## IMSS: Invalidez y Vida ----
  Invalidez <- reactive({
    
    ## Obtengo la informacion de las retenciones del IMSS
    Retenciones_IMSS <- IMSS %>%
      filter(Enfermedades_Maternidad2 == "Invalidez_y_Vida") %>%
      select(Empleado)
    
    DM <- as.numeric(DiasMes())
    
    Retencion <- Retenciones_IMSS/100 * SDI() * DM
    
  })

## IMSS: Cesantia y Vejez ----
  Cesantia <- reactive({
    
    ## Obtengo la informacion de las retenciones del IMSS
    Retenciones_IMSS <- IMSS %>%
      filter(Enfermedades_Maternidad2 == "Cesantia_y_vejez") %>%
      select(Empleado)
    
    DM <- as.numeric(DiasMes())
    
    Retencion <- Retenciones_IMSS/100 * SDI() * DM
    
  })
  
## IMSS: IMSS Mensual ----
  IMSSMens <- reactive({
    
    if (Espec_Exed() + Prestaciones() + Pensiones() + Invalidez() > 78.3) 
      {dValue <- Espec_Exed() + Prestaciones() + Pensiones() + Invalidez()}
    else
      {dValue <- 0}

  })
  
## IMSS: Cesantia y Vejez ----
  CesantiaB <- reactive({
    
    if (Cesantia() > 78.3) 
      {dValue <- Cesantia()}
    else
      {dValue <- 0}
    
  })
  
## IMSS: Retencion IMSS Mensual----
  IMSSRetencion <- reactive({
    
    if (Cesantia() > 78.3) 
      {dValue <- CesantiaB()+ IMSSMens()}
    else
      {dValue <- 0}
    
  })
  
## AAA IMSS: Retencion IMSS ----
  output$PruebaRet <- renderTable({
    
    ## Obtengo la informacion de las retenciones del IMSS
    Retenciones_IMSS <- IMSS %>%
      filter(Empleado > 0) %>%
      select(Enfermedades_Maternidad, Enfermedades_Maternidad2, Empleado)
    ## Valido la informacion
    if(SDI() > 3*SMV)
        {Esp_Ex <- Retenciones_IMSS[which("Especie_Excedente"==Retenciones_IMSS[2]),3]}
    else
        {Esp_Ex <- 0}
    
    DM <- as.numeric(DiasMes())
    
    Retenciones_Montos <- matrix(NaN, length(Retenciones_IMSS[,3]))
    
    for(i in 1:length(Retenciones_IMSS[,3]))
      {if (Retenciones_IMSS[i,2]=="Especie_Excedente")
        {Retenciones_Montos[i,1]<- Esp_Ex/100 * (SDI()-(3*SMV)) * DM}
      # {Retenciones_Montos[i,1]<- Esp_Ex}
          else
            {Retenciones_Montos[i,1] <- Retenciones_IMSS[i,3]/100 * SDI() * DM}
            # {Retenciones_Montos[i,1] <- 725.2*DM*Retenciones_IMSS[i,3]/100}#SDI()*DM*Retenciones_IMSS[i,3]}
      }
    
    df <- data.frame(as.data.frame.vector(Retenciones_IMSS[,1]),Retenciones_Montos)
    df
  })
  
## outPut ISR: ISR Anual ----
  output$ISR_Anual <- renderUI({

          ISR_Anual <- (CuotaFija()+Excedente()*Aplicable()/100)
          
          valueBox(value = prettyNum(ISR_Anual,big.mark=",",scientific=FALSE)#Aplica
                   , subtitle = "ISR Anual:"
                   , icon = icon("money")
                   , color = "navy"
                   , width = 12)
    
  })

## outPut ISR: ISR MensQuin ----
  output$ISR_MenQuin <- renderUI({
    
    ISR_Anual <- ((CuotaFija()+Excedente()*Aplicable()/100)/(TipoSal()*12))
    
    if (TipoSal()==1)
      {sLabel <- "ISR Mensual:"}
    else
      {sLabel <- "ISR Quincenal:"}
    
    valueBox(value = prettyNum(ISR_Anual,big.mark=",",scientific=FALSE)#Aplica
             , subtitle = sLabel
             , icon = icon("money")
             , color = "navy"
             , width = 12)
    
  })

## outPut ISR: Limite Inferior ----  
  output$oLimiteInferior <- renderUI({

    valueBox(value = prettyNum(LimInferior(),big.mark=",",scientific=FALSE)
             , subtitle = "Limite Inferior:"
             , icon = icon("dollar")
             , color = "navy"
             , width = 12)
  })
## outPut ISR: Cuota Fija ----  
  output$oCuotaFija <- renderUI({

    valueBox(value = prettyNum(CuotaFija(),big.mark=",",scientific=FALSE)
             , subtitle = "Cuota Fija:"
             , icon = icon("angle-double-down")
             , color = "navy"
             , width = 12)
  })
## outPut ISR: Excedente ----  
  output$oExcedente <- renderUI({
    
    valueBox(value = prettyNum(Excedente(), big.mark=",",scientific=FALSE)
             , subtitle = "Excedente:"
             , icon = icon("level-up")
             , color = "navy"
             , width = 12)
  })

## outPut IMSS: SMV ----
  output$oSMV <- renderText({SMV})

## outPut IMSS: SDI ----
  output$oSDI <- renderUI({

    valueBox(value = prettyNum(formatC(SDI(), digits =2, format = "f"),big.mark=",",scientific=FALSE)
             , subtitle = "SDI:"
             , icon = icon("hand-spock-o")
             , color = "navy"
             , width = 6)
  })
  
## outPut IMSS: Especie - Excedente ----
  output$oEsEx <- renderUI({
    
    dValue <- trunc(Espec_Exed()*100)/100
    
    valueBox(value = prettyNum(dValue, big.mark=",",scientific=FALSE)
             , subtitle = "Especie y Excedente:"
             , icon = icon("hand-spock-o")
             , color = "navy"
             , width = 6)
  })

## outPut IMSS: Prestaciones en Dinero ----
  output$oPrestaciones <- renderUI({
    
    dValue <- trunc(Prestaciones()*100)/100
    
    valueBox(value = prettyNum(dValue,big.mark=",",scientific=FALSE)
             , subtitle = "Prestaciones:"
             , icon = icon("hand-spock-o")
             , color = "navy"
             , width = 6)
  })
  
## outPut IMSS: Pensiones y Beneficiarios ----
  output$oPensiones <- renderUI({
    
    dValue <- trunc(Pensiones()*100)/100
    
    valueBox(value = prettyNum(dValue,big.mark=",",scientific=FALSE)
             , subtitle = "Pensio. y benef.:"
             , icon = icon("hand-spock-o")
             , color = "navy"
             , width = 6)
  })
  
## outPut IMSS: Invalidez y Vida ----
  output$oInvalidez <- renderUI({
    
    dValue <- trunc(Invalidez()*100)/100
    
    valueBox(value = prettyNum(dValue,big.mark=",",scientific=FALSE)
             , subtitle = "Invalidez y vida:"
             , icon = icon("hand-spock-o")
             , color = "navy"
             , width = 6)
  })
  
## outPut IMSS: Cesantia y Vejez ----
  output$oCesantia <- renderUI({
    
    dValue <- trunc(Cesantia()*100)/100
    
    valueBox(value = prettyNum(dValue, big.mark=",",scientific=FALSE)
             , subtitle = "Cesantia y vejez:"
             , icon = icon("hand-spock-o")
             , color = "navy"
             , width = 6)
  })
  
## outPut IMSS: IMSS Mensual ----
  output$oIMSSMens <- renderUI({
    
    dValue <- trunc(IMSSMens()*100)/100
    
    valueBox(value = prettyNum(dValue, big.mark=",",scientific=FALSE)
             , subtitle = "IMSS Mensual:"
             , icon = icon("hand-spock-o")
             , color = "navy"
             , width = 6)
  })
  
## outPut IMSS: Cesantia Bimestral ----
  output$oCesantiaB <- renderUI({
    
    dValue <- trunc(CesantiaB()*100)/100
    
    valueBox(value = prettyNum(dValue, big.mark=",",scientific=FALSE)
             , subtitle = "Cesantia Bimestral:"
             , icon = icon("hand-spock-o")
             , color = "navy"
             , width = 6)
  })

## outPut IMSS: IMSS Retencion ----
  output$oIMSSRet <- renderUI({
    
    dValue <- trunc(IMSSRetencion()*100)/100
    
    valueBox(value = prettyNum(dValue, big.mark=",",scientific=FALSE)
             , subtitle = "Retencion IMSS:"
             , icon = icon("hand-spock-o")
             , color = "navy"
             , width = 12)
  })
  
## outPut IMSS: Dias del Mes ----
  output$DiasMes <- renderText({
    DiasMes()
  })

  ## Pruebas Print ----  
  output$vLimInf <- renderText({SMV})
  output$vCuotaF <- renderText({CuotaFija()})
  output$vExcedente <- renderText({Excedente()})
  output$vISR <- renderText({ISR_Anual <- (CuotaFija()+Excedente()*Aplicable()/100)})

## Dias oficiales de vacaciones ----
  output$oDiasVac <- renderText({
      VPL()
    })

  output$Prueba <- renderPrint({50})
  
  output$Prueba2 <- renderText({sprintf("Salario actual: %d", 50)})
  output$Prueba3 <- renderUI({
                      valueBox(value = 50
                          , subtitle = "Salario actual:"
                          , icon = icon("list")
                          , color = "blue"
                          , width = 12)})
  output$Prueba4 <- renderUI({
                      valueBox(value = 50
                               , subtitle = "Salario actual:"
                               , icon = icon("list")
                               , color = "blue"
                               , width = 12)})
  
  output$Salario <- renderDataTable({
    datatable(tbSalario,colnames = c("LimI", "Cuota", "Porcentaje"))
  # output$Salario2 <- renderDataTable({
  #   tbSalario
  })
  

  # 
  # ISRAnual <- fISRAnual(Variables[1], Variables[2], Variables[3], SalBAnual)
  # 
  # DeduccionesSalario <- c(Variables[1] , SalBAnual - Variables[1], Variables[3]
  #                         , Variables[2], ISRAnual, ISRAnual/12, ISRAnual/24
  #                         , (SalBAnual - ISRAnual)/12,(SalBAnual - ISRAnual)/24)
  # 
  # tbSal <- data.frame(ListaDeducciones, DeduccionesSalario)
  # tbSal
  
  # output$tbSueldo <- DT::renderDataTable(DT::datatable({
  #   SalBAnual <- input$tSalario*input$Salario*12
  # 
  # Variables <- ISR %>%
  #   filter(Limite_Inferior <= SalBAnual, 
  #          Limite_Superior >= SalBAnual) %>%
  #   select(Limite_Inferior,Cuota_fija,Aplicable)
  # 
  # ListaDeducciones <- c("Limite Inferior", "Excedente", "% Excedente"
  #                       , "Cuota Fija", "ISR Anual", "ISR Mensual"
  #                       , "ISR Quincenal", "Salario Neto Mensual"
  #                       , "Salario Neto Quincenal")
  # 
  # ISRAnual <- fISRAnual(Variables[1], Variables[2], Variables[3], SalBAnual)
  # 
  # DeduccionesSalario <- c(Variables[1] , SalBAnual - Variables[1], Variables[3]
  #                         , Variables[2], ISRAnual, ISRAnual/12, ISRAnual/24
  #                         , (SalBAnual - ISRAnual)/12,(SalBAnual - ISRAnual)/24)
  # 
  # tbSal <- data.frame(ListaDeducciones, DeduccionesSalario)
  # }))
})